os/user: add users groups iteration functionality

mjonaitis1 opened the pull request:
Proposal link: https://github.com/golang/go/issues/47907

Go native users/groups iteration functionality.

Go standard library has os/user package which allows to lookup user
or group records via either user/group name  or id. This pull request
extends capabilities of os/user package by introducing iteration
functionality for users and groups.
Users and groups iteration functionality might be useful in cases where 
a full or partial list of all available users/groups is required in an 
application. 

On unix, iteration functionality relies 
on two internal implementations: one that does not use cgo and parses 
contents of /etc/passwd and /etc/group files for users and groups 
respectively, and another one that does use cgo and utilizes POSIX 
compliant libc library routines getpwent and getgrent for users and 
groups respectively. On windows, 
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList
registry key is iterated for SID values, each SID value represents either
a user or a group. 

Tested on Debian GNU/Linux 10 (buster), macOS 11.5.1, freebsd 13.0, 
plan9 fourth edition, windows 10 version 21H1.

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,139 @@
+// usersHelper defines the methods used in users iteration process within iterateUsers. This interface allows to test
+// iterateUsers functionality. iterate_test_fgetent.go file defines test related struct that implements usersHelper.
+type usersHelper interface {
+	// set does any necessary setup steps before starting iteration
```
nit wording:

```
	// set sets up the internal state before iteration
```

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,139 @@
+	// get sequentially returns a passwd structure which is later processed into *User entry
+	get() (*C.struct_passwd, error)
+
+	// end does any necessary finalization steps after iteration is done
```
similar to `set()`

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// iterate users helper
+var iuh usersHelper = iterateUsersHelper{}
```
explain that this is a global variable because the underlying system is using a global state.

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,139 @@
+// iterate users helper
+var iuh usersHelper = iterateUsersHelper{}
+
+// iterateUsers iterates over users database via getgrent library call. If fn returns non nil error, then
```
`s/getgrent library call/getgrent(3)/`, this is the naming convention.

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,42 @@
+func TestIterateUser(t *testing.T) {
+	err := iterateUsers(func(user *User) error {
+		if user.Username == "" && user.Gid == "" && user.Uid == "" && user.HomeDir == "" && user.Name == "" {
+			t.Errorf("parsed user is empty: %+v \n", user)
```
nit: `Errorf` formats the resulting string with a newline correctly (so you don't need the ` \n`).

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,42 @@
+
+	if !errors.Is(err, _stopErr) {
+		t.Errorf("error encoutered while iterating users: %w", err)
+	}
```
you can check for error equality here, that avoids a reflect call in the back. Also, this is a private API and the error will certanly not be wrapped.

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,42 @@
+	})
+
+	if !errors.Is(err, _stopErr) {
+		t.Errorf("error encoutered while iterating groups: %w", err)
```
stutter in error formatting: `t.Errorf` implies an error:
```
		t.Errorf("iterating groups: %w", err)
```

motiejus commented on src/os/user/iterate_cgo_test.go:
```diff
@@ -0,0 +1,76 @@
+
+	iuh = &iterateUsersHelperTest{}
+
+	// Test that we retrieve groups in same order as we defined
```
use passive voice throughout the comments: "test retrieved groups are in the same order as defined".

motiejus commented on src/os/user/iterate_cgo_test.go:
```diff
@@ -0,0 +1,76 @@
+	})
+
+	if err != nil {
+		t.Errorf("error occurred while iterating users: %v", err)
```
stutter

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,42 @@
+	})
+
+	if !errors.Is(err, _stopErr) {
+		t.Errorf("error encoutered while iterating users: %w", err)
```
stutter

motiejus commented on src/os/user/iterate_cgo_test.go:
```diff
@@ -0,0 +1,76 @@
+	// Use testdata fixture
+	igh = &iterateGroupsHelperTest{}
+
+	// Test that we retrieve groups in same order as we defined
```
use passive voice (will not repeat this comment)

motiejus commented on src/os/user/iterate_example_test.go:
```diff
@@ -0,0 +1,54 @@
+	"fmt"
+)
+
+func ExampleIterateUsers() {
```
normally example tests print something and compare output.

motiejus commented on src/os/user/iterate_example_test.go:
```diff
@@ -0,0 +1,54 @@
+		fmt.Printf("error encountered while iterating groups database: %v", err)
+	}
+
+	// Do something with groups slice
```
is this a TODO?

motiejus commented on src/os/user/iterate_example_test.go:
```diff
@@ -0,0 +1,54 @@
+		fmt.Printf("error encountered while iterating users database: %v", err)
+	}
+
+	// Do something with users slice
```
is this a TODO?

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,99 @@
+func userGroupIterator(lineFn lineFunc) error {
+	f, err := os.Open(usersFile)
+	if err != nil {
+		return fmt.Errorf("opening users file: %w", err)
```
`open users file: %w`

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,99 @@
+		return fmt.Errorf("opening users file: %w", err)
+	}
+	defer func() {
+		_ = f.Close()
```
we probably want to propagate errors

motiejus commented on src/os/user/iterate_plan9_test.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	if !reflect.DeepEqual(wantUsers, gotUsers) {
+		t.Errorf("could not parse users correctly: want: %+v got: %+v \n", wantUsers, gotUsers)
```
ditto; remove the ` \n` from the end of the message.

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,99 @@
+				Gid:      parts[0],
+				Username: parts[1],
+				Name:     parts[1],
+				HomeDir:  "/usr/" + parts[1],
```
`/usr/<username>` is strange, even for plan9; I wasn't able to find a reference that this is how homedir is constructed. Can you point to one?

mjonaitis1 replied:
`/usr/<username>` seems to be the default directory when user is logged in. I could not find anything about this either, so I assume that `/usr/<username>`  is indeed the homedir.

mjonaitis1 replied:
Also **$home** env variable is set to `/usr/<username>` on login

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (i *iterateUsersHelperTest) set() {
+	mode := C.CString("r")
+	defer C.free(unsafe.Pointer(mode))
+	f, err := os.Open(_usersTestDataFile)
+	if err != nil {
+		panic(err)
+	}
+	i.f = f
+	var fp *C.FILE
+	C.resetErrno()
+	fp, err = C.fdopen(C.int(f.Fd()), mode)
+	if fp == nil {
+		panic(err)
+	}
+	i.fp = fp
```
can we open this in Go and pass a file descriptor to C? That may make it simpler to read.

mjonaitis1 replied:
This is what is being done at the moment. What do you mean exactly?

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,117 @@
+
+func (i *iterateUsersHelperTest) end() {
+	if i.f != nil {
+		_ = i.f.Close()
```
errors should not be ignored.

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,117 @@
+	result, err := C.fgetpwent(i.fp)
+
+	// fgetpwent returns ENOENT when there are no more records
+	if result == nil && errors.Is(err, syscall.ENOENT) {
```
can a syscall error be wrapped? If not, then we can check for equality (and add a comment why), since it avoids a reflect call.

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (i *iterateGroupsHelperTest) set() {
+	mode := C.CString("r")
+	defer C.free(unsafe.Pointer(mode))
+	f, err := os.Open(_groupsTestDataFile)
+	if err != nil {
+		panic(err)
+	}
+	i.f = f
+	var fp *C.FILE
+	C.resetErrno()
+	fp, err = C.fdopen(C.int(f.Fd()), mode)
+	if fp == nil {
+		panic(err)
+	}
+	i.fp = fp
```
ditto: less mixing between C and Go when possible.

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,117 @@
+		return nil, nil
+	}
+
+	return result, err
```
by convention, either result xor error are nil or zero values. Never both should have non-zero values.

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,117 @@
+
+func (i *iterateGroupsHelperTest) end() {
+	if i.f != nil {
+		_ = i.f.Close()
```
do not ignore errors

motiejus commented on src/os/user/iterate_unix.go:
```diff
@@ -0,0 +1,59 @@
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = readColonFile(f, usersIterator(fn), 6)
+	return err
+}
+
+func iterateGroups(fn NextGroupFunc) error {
+	f, err := os.Open(groupFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
```
not sure what the current convention in the Go stdlib is: are errors from `Close` calls ignored?

mjonaitis1 replied:
yes, `lookup_unix.go` ignores deferred `f.Close()` calls

motiejus commented on src/os/user/iterate_unix.go:
```diff
@@ -0,0 +1,59 @@
+	return func(line []byte) (interface{}, error) {
+		v, _ := matchUserIndexValue("", -1)(line)
+		if u, ok := v.(*User); ok {
+			err := fn(u)
+			if err != nil {
+				return nil, err
+			}
```
```suggestion
			if err := fn(u); err != nil {
				return nil, err
			}
```

motiejus commented on src/os/user/iterate_unix.go:
```diff
@@ -0,0 +1,59 @@
+func groupsIterator(fn NextGroupFunc) lineFunc {
+	return func(line []byte) (interface{}, error) {
+		v, _ := matchGroupIndexValue("", -1)(line)
+		if g, ok := v.(*Group); ok {
+			err := fn(g)
+			if err != nil {
+				return nil, err
+			}
+		}
```
ditto ^^

motiejus commented on src/os/user/iterate_unix_test.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	for i, u := range collectedUsers {
+		if u.Name != users[i].Name ||
+			u.Gid != users[i].Gid ||
+			u.Uid != users[i].Uid ||
+			u.Username != users[i].Username ||
+			u.HomeDir != users[i].HomeDir {
```
structs can be compared.

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+// Since HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList registry key does not contain SIDs
+// of Administrator, Guest, or other default system users, some users or groups might not be provided when using
+// iterateUsers or iterateGroups
+const ProfileListKey = `SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList`
```
is this mean to be exported?

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+		}
+		u, err := newUserFromSid(SID)
+		if err != nil {
+			return nil
```
return err?

mjonaitis1 replied:
Most probably I mistakenly forgot these 2 cases. We should return nil only where we check if SID is not of User type.

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+		SID, err := syscall.StringToSid(sid)
+
+		if err != nil {
+			return nil
```
return err?

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+	return iterateSIDS(func(sid string) error {
+		SID, err := syscall.StringToSid(sid)
+		if err != nil {
+			return nil
```
return err?

motiejus commented on src/os/user/listgroups_solaris.go:
```diff
@@ -0,0 +1,18 @@
+import "fmt"
+
+func listGroups(u *User) ([]string, error) {
+	return nil, fmt.Errorf("user: list groups for %s: not supported on Solaris", u.Username)
```
this error message implies it may be supported for other users; perhaps `not implemented` is enough?

mjonaitis1 replied:
This is something from stdlib, better not touch it I guess?

motiejus commented on src/os/user/read_colon_file.go:
```diff
@@ -0,0 +1,86 @@
+// to fn; in a long line additional fields may be silently discarded.
+//
+// readColonFile can also be used to read /adm/users on plan9.
+func readColonFile(r io.Reader, fn lineFunc, readCols int) (v interface{}, err error) {
```
does it have to move to a separate file? Moving it to a separate file makes it hard to see what's changed between the old & new versions.

mjonaitis1 replied:
Yes, we need it on plan9, because it does not make sense to create another function which does exactly the same thing. Since originally, this function was in `lookup_unix.go`, which also contains lookup implementation for non-cgo unix version, we need a separate file for this.

mjonaitis1 replied:
Move this to different commit

motiejus commented:
It would be also helpful if you listed the exact OS versions you have tested this with, and which commands did you run when testing.

For bonus points, write a small application in Go compatible to `getent passwd` and `getent group` and compare how they perform:

- is the output identical?
- duration.
- memory usage.

motiejus commented on src/internal/syscall/windows/registry/registry_test.go:
```diff
@@ -35,7 +35,11 @@ func TestReadSubKeyNames(t *testing.T) {
 	defer k.Close()
 
-	names, err := k.ReadSubKeyNames()
+	names := make([]string, 0)
```
```
names := []string{}
```

motiejus commented on src/os/user/lookup_unix.go:
```diff
@@ -147,20 +72,21 @@ func findGroupName(name string, r io.Reader) (*Group, error) {
 // returns a *User for a row if that row's has the given value at the
-// given index.
+// given index. If idx is set to -1, matches and returns any valid *User entry.
+// idx should be set to -1 only for iteration process.
```
remove this line.

motiejus replied:
i.e. cannot prescribe. Perhaps explain why this useful, but not mandate iteration.

motiejus commented on src/os/user/iterate.go:
```diff
@@ -0,0 +1,25 @@
+
+// IterateUsers iterates over user entries. For each retrieved *User entry provided NextUserFunc is called.
+//
+// On UNIX, if CGO is enabled, getpwent is used in the underlying implementation. Since getpwent is not thread-safe,
```
convention: `getpwent(3)`

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,149 @@
+//go:build (aix || dragonfly || freebsd || (!android && linux) || netbsd || openbsd || solaris || darwin) && cgo && !osusergo
+// +build aix dragonfly freebsd !android,linux netbsd openbsd solaris darwin
+// +build cgo
+// +build !osusergo
```
these might be deprecated with go 1.17; please check.

mjonaitis1 replied:
`gofmt`automatically syncs both versions, so at the moment there is no need to remove these.

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,149 @@
+// in tests, iterate_test_fgetent.go file provides iterateUsersHelperTest
+// structure which implements usersHelper interface and can substitute
+// default iuh value.
+var iuh usersHelper = iterateUsersHelper{}
```
`userIterator` / `usersIterator` / something else?

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,149 @@
+			return err
+		}
+
+		// User provided non-nil error means that iteration should be terminated
```
move the comment 1 line below?

motiejus commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,149 @@
+// in tests, iterate_test_fgetent.go file provides iterateGroupsHelperTest
+// structure which implements groupsHelper interface and can substitute
+// default igh value.
+var igh groupsHelper = iterateGroupsHelper{}
```
`groupIterator` / `groupsIterator`?

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,43 @@
+	"testing"
+)
+
+// As cgo implementation can not be tested on darwin or bsd based operating
+// systems, since there is no support for fgetpwent/fgetgrent library calls,
+// we will attempt to check if user/group record could be at least retrieved
```
As BSDs (including darwin) do not support fgetpwent/fgetgrent, attempt to check if at least 1 user/group record can be retrieved.

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,43 @@
+
+func TestIterateUser(t *testing.T) {
+	err := iterateUsers(func(user *User) error {
+		if user.Username == "" && user.Gid == "" && user.Uid == "" && user.HomeDir == "" && user.Name == "" {
```
```
if *user == User{} {
  ...
}
```

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,43 @@
+	if err != _stopErr {
+		t.Errorf("iterating users: %w", err)
+	}
+}
```
also check if at least 1 user was returned.

motiejus commented on src/os/user/iterate_cgo_bsd_test.go:
```diff
@@ -0,0 +1,43 @@
+	if err != _stopErr {
+		t.Errorf("iterating groups: %w", err)
+	}
+}
```
ditto

motiejus commented on src/os/user/iterate_cgo_test.go:
```diff
@@ -0,0 +1,77 @@
+// +build !osusergo
+
+package user
+
```
comment (somewhere) that this file is for unix systems that support fgetpwent/fgetgrent.

motiejus commented on src/os/user/iterate_cgo_test.go:
```diff
@@ -0,0 +1,77 @@
+//go:build (aix || dragonfly || (!android && linux) || solaris) && cgo && !osusergo
+// +build aix dragonfly !android,linux solaris
+// +build cgo
+// +build !osusergo
```
check consistency

motiejus commented on src/os/user/iterate_cgo_test.go:
```diff
@@ -0,0 +1,77 @@
+		t.Errorf("iterating users: %v", err)
+	}
+
+	if len(gotUsers) != len(wantsUsers) || !reflect.DeepEqual(wantsUsers, gotUsers) {
+		t.Errorf("could not parse all users correctly")
+	}
```
1. DeepEqual implies identical lengths.
2. If they are different, they should be printed, so the developer can inspect easier why it fails.

motiejus commented on src/os/user/iterate_example_test.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+func ExampleIterateUsers() {
+	// Get first 20 users
```
you may need a docstring for this to be correctly rendered in the generated docs.

mjonaitis1 replied:
Comment is not actually needed to correctly render the example 
![image](https://user-images.githubusercontent.com/85163890/130952960-cdf57252-85e0-48d1-9408-4373cd61864a.png)

motiejus commented on src/os/user/iterate_example_test.go:
```diff
@@ -0,0 +1,54 @@
+	// Here users slice can be used to do something with collected users.
+}
+
+func ExampleIterateGroups() {
```
ditto

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+// users and groups file location in plan9
+var (
+	usersFile = "/adm/users"
+)
```
- single-variables don't need a block.
- since this is unconventional (should be a const), explain that this is mutated in tests.

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,101 @@
+	if err != nil {
+		return fmt.Errorf("open users file: %w", err)
+	}
+	defer func() {
+		err = f.Close()
+	}()
```
let's be consistent in catching/non-catching of errors from Close.

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// matchPlan9UserGroup matches valid /adm/users line.
+// On plan9 /adm/users lines are both users and groups,
+//
+// idx should be set to -1 for iteration.
+//
```
fix formatting?

motiejus commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,101 @@
+//
+// According to plan 9 users(6): ids are arbitrary text strings, typically the same as name.
+// In older Plan 9 file servers, ids are small decimal numbers.
+func matchPlan9UserGroup(value string, idx int, returnUser bool) lineFunc {
```
consider not using a boolean for this; let's think about and review other approach(-es).

motiejus commented on src/os/user/iterate_plan9_test.go:
```diff
@@ -0,0 +1,124 @@
+const testUsersFile = "./testdata/plan9/users.txt"
+
+func TestIterateUsers(t *testing.T) {
+	usersFile = testUsersFile
```
rename back the file after the test terminates.

motiejus commented on src/os/user/iterate_plan9_test.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func TestIterateGroups(t *testing.T) {
+	usersFile = testUsersFile
```
ditto

motiejus commented on src/os/user/iterate_plan9_test.go:
```diff
@@ -0,0 +1,124 @@
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(wantGroups, gotGroups) {
+		t.Errorf("could not parse groups correctly")
+	}
```
print groups when this happens

motiejus commented on src/os/user/iterate_plan9_test.go:
```diff
@@ -0,0 +1,124 @@
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(wantUsers, gotUsers) {
+		t.Errorf("could not parse users correctly")
+	}
+}
```
- print users when this happens
- it was probably parsed correctly, but the result was invalid.

motiejus commented on src/os/user/iterate_test_fgetent.go:
```diff
@@ -0,0 +1,96 @@
+
+func (i *iterateUsersHelperTest) get() (*C.struct_passwd, error) {
+	var result *C.struct_passwd
+	// fgetpwent returns ENOENT when there are no more records
```
explain why the comment is there: this return value is undocumented in `fgetpwent(3)`. You may also redirect developers to docs of `fgetpwent_r(3)`, which glibc uses for `fgetpwent`.

motiejus commented on src/os/user/iterate_unix_test.go:
```diff
@@ -0,0 +1,136 @@
+`
+	// Ordered list of correctly parsed groups from testGroupFile
+	var groups = []*Group{
+		{
+			Gid:  "-2",
+			Name: "nobody",
+		},
+		{
+			Gid:  "-1",
+			Name: "nogroup",
+		},
+		{
+			Gid:  "7",
+			Name: "indented",
+		},
+		{
+			Gid:  "2",
+			Name: "kmem",
+		},
```
nit: since these are short, you may move every item to a single line:


```
{Gid:  "-2", Name: "nobody",},
```

motiejus commented on src/os/user/iterate_unix_test.go:
```diff
@@ -0,0 +1,136 @@
+
+	r := strings.NewReader(testGroupFile)
+
+	collectedGroups := make([]*Group, 0, 4)
```
use `len(groups)` instead of `4`, since the tests may change, but this value may get out of sync.

motiejus commented on src/os/user/iterate_unix_test.go:
```diff
@@ -0,0 +1,136 @@
+		},
+	}
+
+	collectedUsers := make([]*User, 0, 4)
```
ditto

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+// Since HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList registry key does not contain SIDs
+// of Administrator, Guest, or other default system users, some users or groups might not be provided when using
+// iterateUsers or iterateGroups
+const profileListKey = `SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList`
```
prefix with underscore

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+	"syscall"
+)
+
+// profileListKey registry key contains all local user/group SIDs as sub keys. It is a sub key of HKEY_LOCAL_MACHINE.
```
explain what is SID (in the first occurrence)

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+		}
+
+		// Callback to user supplied fn, with user
+		if err := fn(u); err != nil {
+			return err
+		}
+
+		return nil
```
`return fn(u)`

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+			return err
+		}
+		// Skip non-group SID
+		if t != syscall.SidTypeGroup && t != syscall.SidTypeWellKnownGroup && t != syscall.SidTypeAlias {
+			return nil
+		}
```
can this be combined with the existing `lookupGroupName`?

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+		}
+		g := &Group{Name: groupname, Gid: sid}
+
+		// Callback to user supplied fn, with group
+		if err := fn(g); err != nil {
+			return err
+		}
+
+		return nil
```
ditto shorten

motiejus commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,77 @@
+package user
+
+import (
+	"internal/syscall/windows/registry"
```
let's add a minimal smoke test that runs on windows.

motiejus commented on src/os/user/listgroups_solaris.go:
```diff
@@ -0,0 +1,18 @@
+// Copyright 2016 The Go Authors. All rights reserved.
```
review/remove this file?

motiejus commented on src/os/user/user.go:
```diff
@@ -3,16 +3,23 @@
 resolving user and group ids to names. One is written in pure Go and
 parses /etc/passwd and /etc/group. The other is cgo-based and relies on
 the standard C library (libc) routines such as getpwuid_r and getgrnam_r.
+Iteration functionality, when use with cgo, relies on getpwent and
+getgrent for users and groups respectively.
```
add `(3)`

motiejus commented on src/os/user/user.go:
```diff
@@ -3,16 +3,23 @@
+
+On Windows, iterateUsers and iterateGroups iterates over
+HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProfileList
+registry key to get SIDs of users and groups. Some system users like
```
translate `SID` here too.

motiejus reviewed (commented):
like

Mantelijo committed:
Change ReadSubKeyNames signature to iterator

In order to add users and groups iterators for os/user,
ReadSubKeyNames must be changed. Instead of returning
a full slice of all subkey names, it can be changed to
accept callback function, which receives sequentially
iterated subkey names. As such, ReadSubKeyNames becomes
an iterator and will use less memory for larger sets of
data.

Mantelijo committed:
Move readColonFile to separate file

Plan9 operating system stores users and groups information in
/adm/users file. This file is similar in strucuture to unix
based /etc/passwd and /etc/group. Since readColonFile resides in
lookup_unix.go file, extracting it makes it possible to reuse the
functionality across all os targets that can can utilize the function.

Mantelijo committed:
Add users/groups iteration functionality to os/user

Go standard library has os/user package which allows to lookup user
or group records via either user/group name or id. This commit
extends capabilities of os/user package by introducing iteration
functionality for users and groups.

Users and groups iteration functionality might be useful in cases where
a full or partial list of all available users/groups is required in an
application.

witriew commented on src/internal/syscall/windows/registry/key.go:
```diff
@@ -110,11 +110,15 @@ loopItems:
+			return err
+		}
+
+		// Callback with key name string
```
Comment probably not needed.
```suggestion
```

witriew commented on src/internal/syscall/windows/registry/key.go:
```diff
@@ -87,9 +87,9 @@ func OpenKey(k Key, path string, access uint32) (Key, error) {
-	names := make([]string, 0)
+// ReadSubKeyNames iterates over the names of subkeys of key k. Callback function fn receives each iterated subkey name.
+// If fn returns non-nil error, iteration is terminated and that error is returned.
+func (k Key) ReadSubKeyNames(fn func(string) error) error {
```
The callback style is probably unnecessarily complex.  There doesn't seem to be much gained here.

How large could the list of users be that this would need optimization?

witriew commented on src/os/user/iterate_windows.go:
```diff
@@ -0,0 +1,74 @@
+// iterateSIDS iterates through _profileListKey sub keys and calls provided
+// fn with enumerated sub key name as parameter. If fn returns non-nil error,
+// iteration is terminated.
+func iterateSIDS(fn func(string) error) error {
```
`iterateSIDs` should probably accept a callback that accepts the SID.

```
func iterateSIDs(fn func(*syscall.SID) error) error
```

witriew commented on src/os/user/iterate.go:
```diff
@@ -0,0 +1,25 @@
+// IterateUsers iterates over user entries. For each retrieved *User entry provided NextUserFunc is called.
+//
+// On UNIX, if CGO is enabled, getpwent(3) is used in the underlying implementation. Since getpwent(3) is not thread-safe,
+// locking is strongly advised.
```
What needs locking in this case?  Maybe provide some examples or checks to make sure it is locked?

witriew commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,155 @@
+*/
+import "C"
+
+// usersHelper defines the methods used in users iteration process within
+// iterateUsers. This interface allows testing iterateUsers functionality.
+// iterate_test_fgetent.go file defines test related struct that implements
+// usersHelper.
+type usersHelper interface {
```
Avoid using "helper" in type names.  Recommend: "userIterator", "passwdIterator", etc.

witriew commented on src/os/user/iterate_cgo.go:
```diff
@@ -0,0 +1,155 @@
+	for {
+		var result *C.struct_passwd
+		C.resetErrno()
+		result, err := userIterator.get()
+
+		// If result is nil - getpwent iterated through entire users database or there was an error
+		if result == nil {
+			return err
+		}
```
The return of err should be based off of what `err` actually is instead of the value, `result`.
```
		result, err := userIterator.get()

		// If result is nil - getpwent iterated through entire users database or there was an error
		if err != nil {
			return err
		}
		
		...
		if result == nil {
		        ...
		}
```

witriew commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,97 @@
+func iterateUsers(fn NextUserFunc) error {
+	return userGroupIterator(func(line []byte) (interface{}, error) {
+		v, _ := parsePlan9UserGroup(userReturnFn)(line)
+		if user, ok := v.(*User); ok {
+			err := fn(user)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return nil, nil
```
```suggestion
		if user, ok := v.(*User); !ok {
			return nil, nil
		}
		if err := fn(user); err != nil {
			return nil, err
		}
```

witriew commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,97 @@
+		Username: parts[1],
+		Name:     parts[1],
+
+		// There is no clear documentation which directory is set to homedir for user.
+		// However, when a new user is created, when user logs in, $HOME environment
+		// variable is set to /usr/<username> and this is also the login directory.
+		HomeDir: "/usr/" + parts[1],
```
This seems like a broad assumption.  Would it be better to provide no information rather than potentially false information?

witriew commented on src/os/user/iterate_plan9.go:
```diff
@@ -0,0 +1,97 @@
+
+func iterateUsers(fn NextUserFunc) error {
+	return userGroupIterator(func(line []byte) (interface{}, error) {
+		v, _ := parsePlan9UserGroup(userReturnFn)(line)
```
Leave a comment saying that there isn't ever an error thrown from the underlying call.

witriew reviewed (commented):
Does upstream know about these new API proposals?

A `ListUsers` type API might be more amenable than a callback-based API
